File: midonet-api/src/main/java/org/midonet/api/system_data/TraceRequest.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.api.system_data;

import java.net.URI;
import java.util.UUID;
import javax.xml.bind.annotation.XmlRootElement;

import javax.validation.constraints.NotNull;

import org.midonet.api.ResourceUriBuilder;
import org.midonet.api.UriResource;
import org.midonet.api.filter.Condition;

import org.midonet.cluster.data.TraceRequest.DeviceType;

/* Class representing trace info */
@XmlRootElement
public class TraceRequest extends UriResource {

    @NotNull
    private UUID id;

    @NotNull
    private String name;

    @NotNull
    private DeviceType deviceType;

    @NotNull
    private UUID deviceId;

    @NotNull
    private Condition condition;

    private long limit = Long.MAX_VALUE;

    @NotNull
    private boolean enabled;

    public TraceRequest() {
        super();
    }

    public TraceRequest(UUID id, String name, DeviceType deviceType,
                        UUID deviceId, Condition condition,
                        long limit, boolean enabled) {
        super();
        this.id = id;
        this.name = name;
        this.deviceType = deviceType;
        this.deviceId = deviceId;
        this.condition = condition;
        this.limit = limit;
        this.enabled = enabled;
    }

    public TraceRequest(UUID id, String name, DeviceType deviceType,
                        UUID deviceId, Condition condition) {
        this(id, name, deviceType, deviceId, condition, Long.MAX_VALUE, false);
    }

    public TraceRequest(org.midonet.cluster.data.TraceRequest traceRequest) {
        super();
        this.id = traceRequest.getId();
        this.name = traceRequest.getName();
        this.deviceType = traceRequest.getDeviceType();
        this.deviceId = traceRequest.getDeviceId();
        this.condition = new Condition();
        this.condition.setFromCondition(traceRequest.getCondition());
        this.limit = traceRequest.getLimit();
        this.enabled = (traceRequest.getEnabledRule() != null);
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public UUID getId() {
        return id;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void setDeviceType(DeviceType deviceType) {
        this.deviceType = deviceType;
    }

    public DeviceType getDeviceType() {
        return deviceType;
    }

    public void setDeviceId(UUID deviceId) {
        this.deviceId = deviceId;
    }

    public UUID getDeviceId() {
        return deviceId;
    }

    public void setCondition(Condition condition) {
        this.condition = condition;
    }

    public Condition getCondition() {
        return condition;
    }

    public void setLimit(long limit) {
        this.limit = limit;
    }

    public long getLimit() {
        return limit;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public boolean getEnabled() {
        return enabled;
    }

    public org.midonet.cluster.data.TraceRequest toData() {
        org.midonet.midolman.rules.Condition c = condition.makeCondition();
        return new org.midonet.cluster.data.TraceRequest()
            .setId(id)
            .setName(name)
            .setDeviceType(deviceType)
            .setDeviceId(deviceId)
            .setCondition(c)
            .setLimit(limit);
    }

    @Override
    public String toString() {
        return "TraceRequest{name=" + name
            + ", deviceType=" + deviceType
            + ", deviceId=" + deviceId
            + ", condition=" + condition
            + ", enabled=" + enabled + "}";
    }

    /**
     * @return the self URI
     */
    @Override
    public URI getUri() {
        if (getBaseUri() != null && id != null) {
            return ResourceUriBuilder.getTraceRequest(getBaseUri(), id);
        } else {
            return null;
        }
    }
}

File: python-midonetclient/src/midonetclient/tracerequest.py
# Copyright (c) 2015 Midokura SARL, All Rights Reserved.
# All Rights Reserved
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from midonetclient import condition
from midonetclient import resource_base
from midonetclient import vendor_media_type
from midonetclient import vtep_binding

class TraceRequest(resource_base.ResourceBase, condition.ConditionBase):
    media_type = vendor_media_type.APPLICATION_TRACE_REQUEST_JSON

    def __init__(self, uri, dto, auth):
        super(TraceRequest, self).__init__(uri, dto, auth)
        resource_base.ResourceBase.__init__(self, uri, dto, auth)
        condition.ConditionBase.__init__(self, self.condition_dto)

    def condition_dto(self):
        if not self.dto.has_key('condition'):
            self.dto['condition'] = dict()
        return self.dto['condition']

    def get_id(self):
        return self.dto['id']

    # name
    def get_name(self):
        return self.dto['name']

    def set_name(self, name):
        self.dto['name'] = name
        return self

    # port
    def get_port(self):
        if (self.dto['deviceType'] == "PORT"):
            return self.dto['deviceId']
        else:
            raise KeyError("device isn't a PORT")

    def set_port(self, port):
        self.dto['deviceType'] = "PORT"
        self.dto['deviceId'] = port
        return self

    # router
    def get_router(self):
        if (self.dto['deviceType'] == "ROUTER"):
            return self.dto['deviceId']
        else:
            raise KeyError("device isn't a ROUTER")

    def set_router(self, router):
        self.dto['deviceType'] = "ROUTER"
        self.dto['deviceId'] = router
        return self

    # bridge
    def get_bridge(self):
        if (self.dto['deviceType'] == "BRIDGE"):
            return self.dto['deviceId']
        else:
            raise KeyError("device isn't a BRIDGE")

    def set_bridge(self, bridge):
        self.dto['deviceType'] = "BRIDGE"
        self.dto['deviceId'] = bridge
        return self

    # limit
    def get_limit(self):
        return self.dto['limit']

    def set_limit(self, limit):
        self.dto['limit'] = limit
        return self

    # enabled
    def get_enabled(self):
        return self.dto['enabled']

    def set_enabled(self, enabled):
        self.dto['enabled'] = enabled
        return self


